sql/src/main/java/io/crate/planner/consumer/SemiJoins.java

```diff
@@ -99,48 +99,50 @@ QueriedRelation tryRewrite(QueriedRelation rel, TransactionContext transactionCt
         if (rewriteCandidates.isEmpty()) {
             return null;
         }
-        if (rewriteCandidates.size() > 1) {
-            return null;  // TODO: support this as well
-        }
-        Function rewriteCandidate = rewriteCandidates.get(0);
-        SelectSymbol selectSymbol = getSubqueryOrNull(rewriteCandidate.arguments().get(1));
-        assert selectSymbol != null : "rewriteCandidate must contain a selectSymbol";
-
         AnalyzedRelation sourceRel = getSource(rel);
         if (sourceRel == null) {
             return null;
         }
 
-        // Turn Ref(x) back into Field(rel, x); it's required for the MultiSourceSelect structure;
+
+        // Function to turn Ref(x) back into Field(rel, x); it's required for the MultiSourceSelect structure;
         // (a lot of logic that follows in the Planner after the rewrite is based on Fields)
         java.util.function.Function<? super Symbol, ? extends Symbol> refsToFields =
             RefReplacer.replaceRefs(r -> sourceRel.getField(r.ident().columnIdent(), Operation.READ));
-        Symbol joinCondition = makeJoinCondition(rewriteCandidate, sourceRel, selectSymbol.relation());
-
-        removeRewriteCandidatesFromWhere(rel, rewriteCandidate);
-        QuerySpec newQS = rel.querySpec().copyAndReplace(refsToFields);
 
-        // Avoid name clashes if the subquery is on the same relation; e.g.: select * from t1 where x in (select * from t1)
-        QualifiedName subQueryName = selectSymbol.relation().getQualifiedName().withPrefix("S");
+        removeRewriteCandidatesFromWhere(rel, rewriteCandidates);
+        QuerySpec newTopQS = rel.querySpec().copyAndReplace(refsToFields);
 
         // Using MSS instead of TwoTableJoin so that the "fetch-pushdown" logic in the Planner is also applied
         HashMap<QualifiedName, AnalyzedRelation> sources = new LinkedHashMap<>(2);
         sources.put(rel.getQualifiedName(), sourceRel);
-        sources.put(subQueryName, selectSymbol.relation());
+
+        ArrayList<JoinPair> semiJoinPairs = new ArrayList<>();
```

Comment:
should we code against interfaces rather than implementation?

Reply:
In local scope it's okay - it just shouldn't leak outside (become part of the API).
If it's assigned to a concrete class the JIT doesn't have to resolve the implementation as the compiler will generate different bytecode.